registry/backoff.go

```diff
@@ -0,0 +1,108 @@
+package registry
+
+import (
+	"errors"
+	"net/http"
+	"sync"
+	"time"
+)
+
+var (
+	// For testing
+	now = time.Now
+)
+
+type hostBackoffRoundTripper struct {
+	roundTripper http.RoundTripper
+	max          time.Duration
+	backoffs     map[string]*backoff
+	sync.Mutex
+}
+
+// HostRateLimitedRoundTripper is a http.RoundTripper which applies throttling
+// to requests on a per-host*credentials basis.
+// Note: There's no cleanup on host/cred tuples so this will slowly leak.
+//
+// r              -- upstream roundtripper
+// maxBackoff     -- maximum length to backoff to between request attempts
+func HostBackoffRoundTripper(r http.RoundTripper, maxBackoff time.Duration) http.RoundTripper {
+	return &hostBackoffRoundTripper{
+		roundTripper: r,
+		max:          maxBackoff,
+		backoffs:     map[string]*backoff{},
+	}
+}
+
+func (c *hostBackoffRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	key := backoffKey{host: r.URL.Host}
+	if user, pass, ok := r.BasicAuth(); ok {
+		key.user = user
+		key.pass = pass
+	}
+	c.Lock()
+	b, ok := c.backoffs[key]
+	if !ok {
+		b = &backoff{max: c.max}
+		c.backoffs[key] = b
```

Comment:
I'm not sure why you need to store the backoffs.

The simpler and automatically thread safe option, would be to create a new backoff object for each call. Then you don't need any of this map or locking stuff.

Why do you want to store it? And what benefit do you get from reusing it? Does it need to be that complex?